libcudacxx/include/cuda/std/detail/libcxx/include/__string

```diff
@@ -254,27 +257,31 @@ struct _CCCL_TYPE_VISIBILITY_DEFAULT char_traits<char>
 
   _LIBCUDACXX_HIDE_FROM_ABI static _CCCL_CONSTEXPR_CXX14 int
   compare(const char_type* __s1, const char_type* __s2, size_t __n) noexcept;
-  _LIBCUDACXX_HIDE_FROM_ABI static size_t _CCCL_CONSTEXPR_CXX14 length(const char_type* __s) noexcept
+
+  // Define a fallback _CCCL_BUILTIN_STRLEN function using parens to prevent
+  // macro expansion.
+  _LIBCUDACXX_HIDE_FROM_ABI static _CCCL_CONSTEXPR_CXX14 size_t(_CCCL_BUILTIN_STRLEN)(const char_type* __s) noexcept
   {
-#ifdef _CCCL_HAS_NO_BUILTIN_STRLEN
-#  ifdef _CCCL_BUILTIN_IS_CONSTANT_EVALUATED // is_constant_evaluated only exists since GCC 9
-    if (__libcpp_is_constant_evaluated())
-#  endif // defined(_CCCL_BUILTIN_IS_CONSTANT_EVALUATED)
+    size_t __len = 0;
+    for (; !eq(*__s, char(0)); ++__s)
     {
-      size_t __len = 0;
-      for (; !eq(*__s, char(0)); ++__s)
-      {
-        ++__len;
-      }
-      return __len;
+      ++__len;
     }
-#endif // defined(_CCCL_HAS_NO_BUILTIN_STRLEN)
-#if !defined(_CCCL_HAS_NO_BUILTIN_STRLEN) || defined(_CCCL_BUILTIN_IS_CONSTANT_EVALUATED)
-    NV_IF_ELSE_TARGET(NV_IS_DEVICE,
-                      (size_t __len = 0; for (; !eq(*__s, char(0)); ++__s)++ __len; return __len;),
-                      (return __builtin_strlen(__s);))
-#endif // !defined(_CCCL_HAS_NO_BUILTIN_STRLEN) || defined(_CCCL_BUILTIN_IS_CONSTANT_EVALUATED)
+    return __len;
   }
+  _LIBCUDACXX_HIDE_FROM_ABI static size_t _CCCL_CONSTEXPR_CXX14 length(const char_type* __s) noexcept
+  {
+#if defined(_CCCL_COMPILER_GCC) && _CCCL_GCC_VERSION < 130000
+    // absurd workaround for GCC "internal compiler error: in cxx_eval_array_reference"
+    if (__libcpp_is_constant_evaluated())
+      ;
+#endif
+    // On device, this will always use the fallback function above. On host,
+    // this will use __builtin_strlen if it is available; otherwise, it will
+    // call the fallback function above.
+    NV_IF_ELSE_TARGET(NV_IS_DEVICE, (return (_CCCL_BUILTIN_STRLEN) (__s);), (return _CCCL_BUILTIN_STRLEN(__s);))
```

Comment:
What speaks against:
```suggestion
#if defined(_CCCL_BUILTIN_STRLEN)
    NV_IF_ELSE_TARGET(NV_IS_DEVICE, (return (_CCCL_BUILTIN_STRLEN) (__s);), (return __strlen_fallback(__s);))
#else // ^^^ _CCCL_BUILTIN_STRLEN ^^^ / vvv !_CCCL_BUILTIN_STRLEN vvv
    return __strlen_fallback(__s);
#endif // !_CCCL_BUILTIN_STRLEN 
```

Reply:
apart from the fact that it is 500% more code, nothing. ;-)

Reply:
Thats true, but the 500% are about 5 lines and we dont have to fake a macro invocation